dozer-core/src/dag/executor.rs

```diff
@@ -166,43 +167,75 @@ impl<'a, T: Clone + Debug + 'static> DagExecutor<'a, T> {
         })
     }
 
-    pub fn validate(dag: &'a Dag<T>, path: &Path) -> Result<(), ExecutionError> {
-        Self::load_or_init_schema(dag, path).map(|_| ())
+    pub fn validate(dag: &Dag<T>, path: &Path) -> Result<(), ExecutionError> {
+        let dag_schemas = DagSchemas::new(dag)?;
+        let meta_manager = DagMetadataManager::new(dag, path)?;
+
+        let current_schemas = dag_schemas.get_all_schemas();
+        let existing_schemas = meta_manager.get_metadata()?;
+        for (handle, current) in &current_schemas {
+            if let Some(existing) = existing_schemas.get(handle) {
+                Self::validate_schemas(current, existing)?;
+            } else {
+                // Non-existing schemas is OK. `Executor::new` will initialize the schemas.
+            }
+        }
+
+        Ok(())
     }
 
     fn validate_schemas(
         current: &NodeSchemas<T>,
         existing: &DagMetadata,
     ) -> Result<(), ExecutionError> {
         if existing.output_schemas.len() != current.output_schemas.len() {
-            return Err(IncompatibleSchemas());
+            return Err(IncompatibleSchemas(
+                "Output Schemas length mismatch".to_string(),
+            ));
         }
         for (port, (schema, _ctx)) in &current.output_schemas {
             let other_schema = existing
                 .output_schemas
                 .get(port)
-                .ok_or(IncompatibleSchemas())?;
+                .ok_or(IncompatibleSchemas(format!(
+                    "Cannot find output schema on port {port:?}"
+                )))?;
             if schema != other_schema {
-                return Err(IncompatibleSchemas());
+                schema.print().printstd();
+
+                other_schema.print().printstd();
+                return Err(IncompatibleSchemas(format!(
+                    "Schema mismatch for port {port:?}:"
+                )));
             }
         }
         if existing.input_schemas.len() != current.input_schemas.len() {
-            return Err(IncompatibleSchemas());
+            return Err(IncompatibleSchemas(
+                "Input Schemas length mismatch".to_string(),
+            ));
         }
-        for (port, (schema, _ctx)) in &current.output_schemas {
-            let other_schema = existing
-                .output_schemas
-                .get(port)
-                .ok_or(IncompatibleSchemas())?;
+        for (port, (schema, _ctx)) in &current.input_schemas {
```

Comment:
should be `output_schemas`?

Reply:
I'm not sure. @v3g42 said it was a fix.

Reply:
No. I am not sure about this code actually. I saw two code blocks that repeated. I assumed one was for input_schemas. Can you check the logic properly again ?

Reply:
This commit's version is correct. Resovling this conversation.